### DISCUSSION TITLE
Predicting RAM Usage by GLOP Solver from OR-Tools: Inconsistencies Observed

### DISCUSSION THREAD

**Discussion by Diddlik** [google/or-tools#3895] [Linear Solver questions] [answered]
**What version of OR-Tools and what language are you using?**
Version: v8.10  - v9.6 
Language: C++

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
GLOP

**What operating system (Linux, Windows, ...) and version?**
Windows

Hello everyone,

I've been utilizing the GLOP solver from OR-Tools for various optimization tasks in my project. An observation I've made is the inconsistent RAM usage by the solver. For instance, there are instances where it uses approximately twice the memory of the original problem, and in other cases, it uses up to 10 times the memory. On average, the RAM consumption hovers around 4 times that of the initial problem.

While I'm aware that memory usage can vary based on the complexity of the problem, I'm struggling to find a linear correlation between problem size and memory usage. **Being able to predict the potential RAM consumption would be immensely beneficial for resource planning in my work.**

Has anyone else observed such inconsistencies? More importantly, are there insights or techniques available to better anticipate the memory requirements for a given problem?

Your guidance or any pointers on this matter would be of immense value.

Thank you!

**Comment by Mizux** [comment]
does `P space complexity == NP problem complexity` ?

**Reply by Diddlik** [reply]
it seems to be polynomial.

**Comment by lperron** [comment]
I do not see the issue.

The performance of a simplex is completely data dependant. 
I would have never expected a simple rule about running time. Otherwise the Netlib would never have hard and smaller problems.

**Reply by Diddlik** [reply]
the runtime is clear that this depends on the problem, but the different allocation of memory remains unclear.

**Reply by Diddlik** [reply]
if i create a problem which is e.g. 500mb, during the whole calculation the memory requirement increases to e.g. 2GB.
For the other problem, which is again 500mb, memory consumption increases to 4gb.

**Reply by sschnug** [reply, marked as answer]
Even if being restricted to LPs (opposed to MIPs where everything is worse), i would not be surprised to see simplex-related data-dependent memory-effects, but that's a guess an simplex-style solvers are complex.

**Howewer: Solving LPs also means doing presolve. With presolve it's obvious, that the statistics / combinatorial nature of the data determine how big the presolved problem is.**

Most of this is related to *sparsity*. 
Simple example:

- Take some sparse matrix `A` and calculate `A*A.T`. This absoluty kills sparsity and in many cases you cannot even store the result anymore
  - There are some well-known structures where bad behaviour is expected, e.g. dense columns (vs. interior-point solvers)

If you are less interested in this simple semi-related example, read:

> Andersen, Erling D., and Knud D. Andersen. "Presolving in linear programming." Mathematical Programming 71 (1995): 221-245.

Maybe also turn off presolve to see how much of your observation changes (or not).
